### ISSUE TITLE
Maybe switch to Mocha?

### ISSUE THREAD

**Issue by davidtheclark** [stylelint/stylelint#31]
I am generally a fan of tape, and I know @MoOx is, too; but I've been writing tests for `block-opening-brace-before` (neverending ...) and for a few reasons I'm wondering if for this particular situation Mocha would be more useful.

At the bottom of the message I'll paste the test file I have so far, so you can see some context for what I'm talking about. Maybe you know of ways to solve these problems with tape?

- Being able to further organize this test files with *indefinite nesting of `describe()` functions* would be good. 
- Describing the text *before* the assertion also seems in my mind to make such a long test file easier to read.
- The TAP output is pretty unreadable at such length without better nesting (unless maybe we make much longer assertion descriptions than I'm doing?) --- certainly not as nice as if it were organized into nested sections, as in the default Mocha output. Could just one of the "pretty reporters" (https://github.com/substack/tape#pretty-reporters).
- Mocha's async handling with callbacks might provide us better means to write test utility functions that tape's `plan()`. For example, I imagine writing utility functions for `expectSuccess()` or (no warnings) or `expectFailure()` (check for warning and correct message), and while I *can* think of a straightforward way to do this with callbacks and unlimited nesting of `describe()`, I'm having some trouble thinking of such a way with `plan()` and not-unlimited-nesting. (Also, callbacks might get around the need to change the number of planned tests whenever you add or remove tests, which is a big annoyance.)

Those are my concerns. What do you think? Can we solve them with tape? Should we switch to Mocha?

I partly feel like the only reason *not* to switch to tape (unless these concerns are unfounded) would be some ideal of purity which we should probably ignore in favor of pragmatism... Maybe it's worth an experiment, at least.

Can anybody foresee problems with Mocha?

```js
import test from "tape"
import testRule from "../../../testUtils/testRule"

import blockOpeningBraceBefore from ".."
import { messages } from ".."

const testRule = testRule(blockOpeningBraceBefore)

test("block-opening-brace-before success", t => {
  t.test("with `space`", st => {
    st.plan(4)

    testRule("body { color: pink; }", "space", warnings => {
      st.equal(warnings.length, 0, "single-line rule")
    })

    testRule("body {\n\tcolor: pink;\n}", "space", warnings => {
      st.equal(warnings.length, 0, "multi-line rule")
    })

    testRule("@media print { color: pink; }", "space", warnings => {
      st.equal(warnings.length, 0, "single-line at-rule")
    })

    testRule("@media print {\n\tcolor: pink;\n}", "space", warnings => {
      st.equal(warnings.length, 0, "multi-line at-rule")
    })
  })

  t.test("with `newline`", st => {
    st.plan(2)

    testRule("body\n{ color: pink; }", "newline", warnings => {
      st.equal(warnings.length, 0, "multi-line rule")
    })

    testRule("@media print\n{\n\tcolor: pink;\n}", "newline", warnings => {
      st.equal(warnings.length, 0, "multi-line at-rule")
    })
  })

  t.test("with `{ singleLine: `space`, multiLine: `space` }`", st => {
    const options = { singleLine: "space", multiLine: "space" }
    st.plan(4)

    testRule("body { color: pink; }", options, warnings => {
      st.equal(warnings.length, 0, "single-line rule")
    })

    testRule("body {\n\tcolor: pink;\n}", options, warnings => {
      st.equal(warnings.length, 0, "multi-line rule")
    })

    testRule("@media print { color: pink; }", options, warnings => {
      st.equal(warnings.length, 0, "single-line at-rule")
    })

    testRule("@media print {\n\tcolor: pink;\n}", options, warnings => {
      st.equal(warnings.length, 0, "multi-line at-rule")
    })
  })

  t.test("with `{ singleLine: `space`, multiLine: `newline` }`", st => {
    const options = { singleLine: "space", multiLine: "newline" }
    st.plan(4)

    testRule("body { color: pink; }", options, warnings => {
      st.equal(warnings.length, 0, "single-line rule")
    })

    testRule("body\n{\n\tcolor: pink;\n}", options, warnings => {
      st.equal(warnings.length, 0, "multi-line rule")
    })

    testRule("@media print { color: pink; }", options, warnings => {
      st.equal(warnings.length, 0, "single-line at-rule")
    })

    testRule("@media print\n{\n\tcolor: pink;\n}", options, warnings => {
      st.equal(warnings.length, 0, "multi-line at-rule")
    })
  })

  t.end()
})

test.only("block-opening-brace-before failure", t => {
  t.test("warns when there are no whitespaces before opening brace", st => {
    st.plan(8)

    testRule("body{ color: pink; }", "space", warnings => {
      st.equal(warnings.length, 1, "with `space` setting and rule")
      st.equal(warnings[0].text, messages.expected("space"), "warning contains expected test")
    })

    testRule("body{ color: pink; }", "newline", warnings => {
      st.equal(warnings.length, 1, "with `newline` setting and rule")
      st.equal(warnings[0].text, messages.expected("newline"), "warning contains expected test")
    })

    testRule("@media print{ color: pink; }", "space", warnings => {
      st.equal(warnings.length, 1, "with `space` setting and at-rule")
      st.equal(warnings[0].text, messages.expected("space"), "warning contains expected test")
    })

    testRule("@media print{ color: pink; }", "newline", warnings => {
      st.equal(warnings.length, 1, "with `newline` setting and at-rule")
      st.equal(warnings[0].text, messages.expected("newline"), "warning contains expected test")
    })
  })

  t.test("when there are multiple whitespaces before opening brace", st => {
    st.plan(8)

    testRule("body  { color: pink; }", "space", warnings => {
      st.equal(warnings.length, 1, "warns with two spaces before brace of rule")
      st.equal(warnings[0].text, messages.expected("space"), "warning contains expected test")
    })

    testRule("body\n\t{ color: pink; }", "newline", warnings => {
      st.equal(warnings.length, 1, "warns with \\n\\t before brace of rule")
      st.equal(warnings[0].text, messages.expected("newline"), "warning contains expected test")
    })

    testRule("@media print  { color: pink; }", "space", warnings => {
      st.equal(warnings.length, 1, "warns with two spaces before brace of at-rule")
      st.equal(warnings[0].text, messages.expected("space"), "warning contains expected test")
    })

    testRule("@media print\n\t{ color: pink; }", "newline", warnings => {
      st.equal(warnings.length, 1, "warns with \\n\\t before brace of at-rule")
      st.equal(warnings[0].text, messages.expected("newline"), "warning contains expected test")
    })
  })

  t.test("with `space`", st => {
    st.plan(4)

    testRule("body\n{ color: pink; }", "space", warnings => {
      st.equal(warnings.length, 1, "warns if actual for rule is newline")
      st.equal(warnings[0].text, messages.expected("space"), "warning contains expected test")
    })

    testRule("@media print\n{ color: pink; }", "space", warnings => {
      st.equal(warnings.length, 1, "warns if actual for at-rule is newline")
      st.equal(warnings[0].text, messages.expected("space"), "warning contains expected test")
    })
  })

  t.test("with `newline`", st => {
    st.plan(4)

    testRule("body { color: pink; }", "newline", warnings => {
      st.equal(warnings.length, 1, "warns if actual for rule is space")
      st.equal(warnings[0].text, messages.expected("newline"), "warning contains expected test")
    })

    testRule("@media print { color: pink; }", "newline", warnings => {
      st.equal(warnings.length, 1, "warns if actual for at-rule is space")
      st.equal(warnings[0].text, messages.expected("newline"), "warning contains expected test")
    })
  })

  t.test("with `{ singleLine: `space`, multiLine: `space` }`", st => {
    const options = { singleLine: "space", multiLine: "space" }
    st.plan(4)

    testRule("body\n{\ncolor: pink; }", options, warnings => {
      st.equal(warnings.length, 1, "warns if newline precedes multiline rule")
      st.equal(warnings[0].text, messages.expected("space"), "warning contains expected test")
    })

    testRule("@media print\n{\ncolor: pink; }", options, warnings => {
      st.equal(warnings.length, 1, "warns if newline precedes multiline at-rule")
      st.equal(warnings[0].text, messages.expected("space"), "warning contains expected test")
    })
  })

  t.end()
})
```

**Comment by davidtheclark** [comment]
I tried https://github.com/substack/faucet, and it helps errors stand out but does not have clear sectioning & nesting.

Also tried https://github.com/scottcorgan/tap-spec and it is good and has better one-level sectioning but doesn't show two-level well (which is the most that tape can do).

**Comment by davidtheclark** [comment]
Maybe I should try to nest less and use more `test()` functions ...

**Comment by MoOx** [comment]
First of all, I think your test file is too big. You should use smaller files. This makes easy to work specific thing by testing the small file only, and for me it makes the output problem away (that said I locally pipe sometimes using tap-spec)

With small files, you can indeed use moooore test() function directly so description is above.

You can also avoid plan() & use end() instead

Here are latests tests I wrote with tape https://github.com/FWeinb/metalsmith-watch/tree/master/__tests__

Tell me what you think.

**Comment by davidtheclark** [comment]
I'm not sure the file size is part of the problem; but I do agree that I should split this one up. I could just use more `test()` functions in a large file. More files != more clear output, right? -- because filename is not part of the output. But smaller test files would make the files a little easier to read, so it's a start.

I guess you're right that I can use `end()` here ... I assumed I would not be able to because PostCSS is all async with its promises etc. But if I switch out out to `end()` it does still pass. Maybe I misunderstand the async of PostCSS? Well, either way that's good: that makes things simpler.

I'll give it another go ...

**Comment by davidtheclark** [comment]
Actually I'm not so sure `end()` did not just work coincidentally ...

**Comment by davidtheclark** [comment]
I'm satisfied staying with Mocha and think the bigger question is how to speed up test writing, which we can discuss here: https://github.com/stylelint/stylelint/issues/32
